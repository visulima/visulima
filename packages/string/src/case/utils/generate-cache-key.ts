import type { CaseOptions } from "../types";

/**
 * Generates a fast cache key from the input string and relevant options.
 * This is optimized for speed using string concatenation which is highly optimized in V8.
 */
const generateCacheKey = (value: string, options?: CaseOptions & { joiner?: string }): string =>
    // Only include options that affect the output
    `${value}::${options?.joiner ?? ""}::${options?.locale ?? ""}::${options?.knownAcronyms?.join(",") ?? ""}::${options?.normalize ?? false}`;

export default generateCacheKey;
