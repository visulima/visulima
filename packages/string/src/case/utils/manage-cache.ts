/**
 * Manages cache operations for case conversion functions.
 * Handles cache size limits by removing the oldest entries when needed.
 */
export const manageCache = (cacheStore: Map<string, string>, key: string, value: string, maxSize: number): void => {
    // If we're at the size limit, remove the first (oldest) entry
    if (cacheStore.size >= maxSize) {
        const firstKey = cacheStore.keys().next().value;

        cacheStore.delete(firstKey);
    }

    // Add the new value
    cacheStore.set(key, value);
};
