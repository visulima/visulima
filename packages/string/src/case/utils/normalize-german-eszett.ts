/**
 * Converts German "SS" to "ß" when appropriate for German locales.
 * This is used to handle special cases in German text where "SS" should be converted to "ß".
 * @param value The string to process
 * @returns The processed string with appropriate eszett conversions
 * @example
 * ```typescript
 * normalizeGermanEszett("GROSSE") // => "GROßE"
 * ```
 */
const normalizeGermanEszett = (value: string): string =>
    // Replace "SS" with "ß" only when both letters are uppercase
    // This uses a positive lookbehind to ensure we don't convert "ss" or "Ss"
    value.replaceAll(/(?<![a-zß])SS(?![a-z])/g, "ß");

export default normalizeGermanEszett;
