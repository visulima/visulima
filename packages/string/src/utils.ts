import type { Interval, IntervalArray } from "./types";

 
export { default as LRUCache } from "./utils/lru-cache";

/**
 * Escapes characters with special meaning in regular expressions.
 * @param str The string to escape.
 * @returns The escaped string.
 */
export function escapeRegExp(string_: string): string {
    // $& means the whole matched string
    return string_.replaceAll(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * Checks if a number falls within any of the specified intervals.
 * @param num The number to check.
 * @param ranges An array of intervals [start, end].
 * @returns True if the number is within any interval, false otherwise.
 */
export function inRange(number_: number, ranges: IntervalArray): boolean {
    return ranges.some((range) => number_ >= range[0] && number_ <= range[1]);
}

/**
 * Checks if a string contains Chinese characters (Han script).
 * @param str The string to check.
 * @returns True if the string contains Chinese characters, false otherwise.
 */
export function hasChinese(string_: string): boolean {
    // Uses Unicode property escapes for Han script
    return /\p{Script=Han}/u.test(string_);
}

/**
 * Checks if a string contains punctuation or space characters.
 * @param str The string to check.
 * @returns True if the string contains punctuation or space, false otherwise.
 */
export function hasPunctuationOrSpace(string_: string): boolean {
    // Uses Unicode property escapes for Punctuation and general category for Space
    return /[\p{P}\p{Z}]/u.test(string_);
}

/**
 * Finds all occurrences of a set of substrings within a source string and returns their start/end indices.
 * @param source The string to search within.
 * @param needles An array of strings to search for.
 * @returns An array of intervals [start, end] for each occurrence.
 */
export function findStrOccurrences(source: string, needles: string[]): IntervalArray {
    const ranges: IntervalArray = [];
    if (!needles || needles.length === 0) {
        return ranges;
    }

    needles.forEach((needle) => {
        if (typeof needle !== "string" || needle.length === 0) {
            return; // Skip invalid needles
        }
        let startIndex = 0;
        let index;
        while ((index = source.indexOf(needle, startIndex)) > -1) {
            const end = index + needle.length - 1;
            ranges.push([index, end]);
            startIndex = index + 1; // Move past the start of the current match
        }
    });

    // Sort and merge overlapping/adjacent intervals for cleaner output
    if (ranges.length === 0) {
        return [];
    }

    ranges.sort((a, b) => a[0] - b[0]);

    const merged: IntervalArray = [];
    let currentRange: Interval | null = null;

    for (const range of ranges) {
        if (currentRange === null) {
            currentRange = [...range];
        } else if (range[0] <= currentRange[1] + 1) {
            // Merge overlapping or adjacent ranges
            currentRange[1] = Math.max(currentRange[1], range[1]);
        } else {
            // Disjoint range, push the previous one and start a new one
            merged.push(currentRange);
            currentRange = [...range];
        }
    }

    if (currentRange !== null) {
        merged.push(currentRange);
    }

    return merged;
}

/**
 * Represents a potential replacement match.
 */
interface PotentialMatch {
    start: number;
    end: number;
    replacement: string;
    original: string;
}

/**
 * Search and replace a list of strings/regexps respecting ignore ranges.
 * Builds the string segment by segment, applying only non-overlapping, non-ignored replacements.
 */
export function replaceString(source: string, searches: OptionReplaceArray, ignoreRanges: IntervalArray): string {
    // console.log(`[replaceString] Initial Source: "${source}", Ignores: ${JSON.stringify(ignoreRanges)}`); // Keep log for now
    const potentialMatches: PotentialMatch[] = [];

    // 1. Find all potential matches from the original source string
    for (const item of searches) {
        if (!item || item.length < 2) continue;
        const [searchKey, replacementValue] = item;
        if (replacementValue === undefined) continue;

        let searchPattern: RegExp;
        try {
             if (searchKey instanceof RegExp) {
                // Ensure the regex has the global flag for exec to work correctly in a loop
                searchPattern = new RegExp(searchKey.source, `${searchKey.flags.replace("g", "")}g`);
            } else if (typeof searchKey === "string" && searchKey.length > 0) {
                searchPattern = new RegExp(escapeRegExp(searchKey), "g");
            } else {
                continue; // Skip invalid search keys
            }

            let match;
            while ((match = searchPattern.exec(source)) !== null) {
                const start = match.index;
                const end = start + match[0].length - 1;
                const original = match[0];

                // Handle $ replacements in the replacement string here
                const finalReplacement = replacementValue.replace(/\$(\d+|&)/g, (_, group) => {
                    if (group === '&') return original;
                    const groupIndex = parseInt(group, 10);
                    return groupIndex > 0 && groupIndex < match.length ? match[groupIndex] ?? '' : '';
                });

                potentialMatches.push({ start, end, replacement: finalReplacement, original });

                // Prevent infinite loops for zero-length matches
                if (match[0].length === 0) {
                     if (searchPattern.lastIndex >= source.length) break;
                    searchPattern.lastIndex++;
                }
            }
        } catch (error) {
             console.error(`Error processing search key ${String(searchKey)}:`, error);
             continue;
        }
    }

    // 2. Sort potential matches by start index, then by length descending (longer matches take precedence)
    potentialMatches.sort((a, b) => {
        if (a.start !== b.start) {
            return a.start - b.start;
        }
        return b.end - a.end; // Longer match first if starts are equal
    });

    // 3. Build the result string, applying non-overlapping, non-ignored matches
    let result = "";
    let lastIndex = 0;
    const appliedRanges: IntervalArray = []; // Keep track of where replacements were made

    for (const match of potentialMatches) {
        // Check for overlap with already applied replacements
        const overlapsApplied = appliedRanges.some(
            (applied) => match.start <= applied[1] && match.end >= applied[0]
        );
        if (overlapsApplied) {
            continue; // Skip this match if it overlaps with one already applied
        }

        // Check for overlap with ignored ranges
        const isIgnored = ignoreRanges.some(
            (ignored) => match.start <= ignored[1] && match.end >= ignored[0]
        );
        if (isIgnored) {
            continue; // Skip this match if it overlaps with an ignored range
        }

        // If we reach here, the match is valid and should be applied
        // Append the text from the last index up to the start of this match
        if (match.start > lastIndex) {
             result += source.substring(lastIndex, match.start);
        }
        // Append the replacement
        result += match.replacement;
        // Update the last index to the end of this match
        lastIndex = match.end + 1;
        // Record the range that was replaced
        appliedRanges.push([match.start, match.end]);
    }

    // 4. Append any remaining part of the source string
    if (lastIndex < source.length) {
        result += source.substring(lastIndex);
    }

    return result;
}
