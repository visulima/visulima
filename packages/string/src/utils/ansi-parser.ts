import { ANSI_ESCAPE_BELL, ANSI_ESCAPE_LINK, ANSI_SGR_TERMINATOR, ESCAPES } from "../constants";
import AnsiStateTracker from "./ansi-state-tracker";
import type { AnsiSegment, HyperlinkSegment, ProcessAnsiStringOptions } from "./types";

/**
 * Helper function to check if a character is inside an ANSI escape sequence
 * @param chars - Array of characters
 * @param index - Current index
 * @returns Object with isInsideEscape and isInsideLinkEscape flags
 */
export const checkEscapeSequence = (
    chars: string[],
    index: number,
): {
    isInsideEscape: boolean;
    isInsideLinkEscape: boolean;
} => {
    if (!ESCAPES.has(chars[index] as string)) {
        return { isInsideEscape: false, isInsideLinkEscape: false };
    }

    const isInsideEscape = true;
    // Check for link escape sequence
    const possibleLink = chars.slice(index + 1, index + 1 + ANSI_ESCAPE_LINK.length).join("");
    const isInsideLinkEscape = possibleLink === ANSI_ESCAPE_LINK;

    return { isInsideEscape, isInsideLinkEscape };
};

/**
 * Process a string with ANSI escape codes character by character
 * @param string - The string to process
 * @param options - Processing options
 */
// eslint-disable-next-line sonarjs/cognitive-complexity
export const processAnsiString = (string: string, options: ProcessAnsiStringOptions = {}): void => {
    const stateTracker = new AnsiStateTracker();

    let currentText = "";
    let isInsideEscape = false;
    let isInsideLinkEscape = false;
    let escapeBuffer = "";
    let currentUrl = "";
    let isInHyperlink = false;
    let textIndex = 0;
    let visualIndex = 0;

    const chars = [...string];

    // eslint-disable-next-line no-plusplus,no-loops/no-loops
    for (let index = 0; index < chars.length; index++) {
        const character = chars[index] as string;

        // Handle escape sequences
        if (character && ESCAPES.has(character)) {
            // If we have pending text, emit it as a segment
            if (currentText) {
                const width = options.getWidth?.(currentText) ?? 0;
                const textStart = textIndex - currentText.length;
                const textEnd = textIndex;

                const segment: AnsiSegment | HyperlinkSegment = {
                    isEscapeSequence: false,
                    isGrapheme: true,
                    text: currentText,
                    textEnd,
                    textStart,
                    visualEnd: visualIndex,
                    visualStart: visualIndex - width,
                    width,
                };

                // If we're inside a hyperlink, add the link info to the segment
                if (isInHyperlink) {
                    (segment as HyperlinkSegment).isHyperlink = true;
                    (segment as HyperlinkSegment).hyperlinkUrl = currentUrl;
                }

                if (options.onSegment?.(segment, stateTracker) === false) {
                    return;
                }

                currentText = "";
            }

            isInsideEscape = true;
            escapeBuffer = character;
            // eslint-disable-next-line no-plusplus
            textIndex++;

            // Check for hyperlink sequence
            const escapeInfo = checkEscapeSequence(chars, index);
            isInsideLinkEscape = escapeInfo.isInsideLinkEscape;

            if (isInsideLinkEscape) {
                // Extract URL from hyperlink sequence
                let urlEnd = index + 1;
                currentUrl = "";

                // eslint-disable-next-line no-loops/no-loops
                while (urlEnd < chars.length) {
                    const nextChar = chars[urlEnd] as string;

                    if (nextChar === ANSI_ESCAPE_BELL) {
                        break;
                    }

                    currentUrl += nextChar;
                    // eslint-disable-next-line no-plusplus
                    urlEnd++;
                    // eslint-disable-next-line no-plusplus
                    textIndex++;
                }

                // Remove the "]8;;" prefix
                currentUrl = currentUrl.slice(4);

                const segment: HyperlinkSegment = {
                    hyperlinkUrl: currentUrl,
                    isEscapeSequence: true,
                    isGrapheme: false,
                    isHyperlink: true,
                    isHyperlinkStart: true,
                    textEnd: textIndex + 1, // +1 to include the bell char
                    textStart: textIndex - escapeBuffer.length - currentUrl.length - 1, // -1 for bell char
                    visualEnd: visualIndex,
                    visualStart: visualIndex,
                    width: 0,
                };

                if (options.onSegment?.(segment, stateTracker) === false) {
                    return;
                }

                index = urlEnd;
                isInHyperlink = true;
                isInsideEscape = false;
                isInsideLinkEscape = false;
                escapeBuffer = "";
                // eslint-disable-next-line no-plusplus
                textIndex++; // For the bell character

                // eslint-disable-next-line no-continue
                continue;
            }

            // Check for hyperlink end sequence: \u001B\\
            if (index + 1 < chars.length && chars[index + 1] === "\\" && isInHyperlink) {
                const segment: HyperlinkSegment = {
                    isEscapeSequence: true,
                    isGrapheme: false,
                    isHyperlink: true,
                    isHyperlinkEnd: true,
                    textEnd: textIndex + 2, // +2 for escape and backslash
                    textStart: textIndex,
                    visualEnd: visualIndex,
                    visualStart: visualIndex,
                    width: 0,
                };

                if (options.onSegment?.(segment, stateTracker) === false) {
                    return;
                }

                isInHyperlink = false;
                currentUrl = "";
                // eslint-disable-next-line no-plusplus
                index++; // Skip the backslash
                // eslint-disable-next-line no-plusplus
                textIndex++; // For the backslash
                isInsideEscape = false;
                escapeBuffer = "";

                // eslint-disable-next-line no-continue
                continue;
            }
        }

        if (isInsideEscape) {
            escapeBuffer += character;
            // eslint-disable-next-line no-plusplus
            textIndex++;

            if (character === ANSI_SGR_TERMINATOR) {
                // End of SGR sequence
                isInsideEscape = false;
                stateTracker.processEscape(escapeBuffer);

                // Determine style information
                // eslint-disable-next-line no-control-regex,regexp/no-control-character
                const styleType = /\u001B\[(\d+)m/.exec(escapeBuffer)?.[1];

                let isCloseStyle = false;
                let isOpenStyle = false;

                if (styleType) {
                    isOpenStyle = !["0", "39", "49"].includes(styleType) && !(Number.parseInt(styleType, 10) >= 21 && Number.parseInt(styleType, 10) <= 29);
                    isCloseStyle = ["0", "39", "49"].includes(styleType) || (Number.parseInt(styleType, 10) >= 21 && Number.parseInt(styleType, 10) <= 29);
                }

                const segment: AnsiSegment = {
                    isCloseStyle,
                    isEscapeSequence: true,
                    isGrapheme: false,
                    isOpenStyle,
                    styleType,
                    text: escapeBuffer,
                    textEnd: textIndex,
                    textStart: textIndex - escapeBuffer.length,
                    visualEnd: visualIndex,
                    visualStart: visualIndex,
                    width: 0,
                };

                if (options.onSegment?.(segment, stateTracker) === false) {
                    return;
                }

                escapeBuffer = "";
            }

            // eslint-disable-next-line no-continue
            continue;
        }

        // Accumulate regular characters
        currentText += character;
        // eslint-disable-next-line no-plusplus
        textIndex++;

        // For single-character segments, emit each one separately
        const width = options.getWidth?.(currentText) ?? currentText.length;
        visualIndex += width;
    }

    // Add any remaining text
    if (currentText) {
        const width = options.getWidth?.(currentText) ?? 0;
        const textStart = textIndex - currentText.length;
        const textEnd = textIndex;

        const segment: AnsiSegment | HyperlinkSegment = {
            isEscapeSequence: false,
            isGrapheme: true,
            text: currentText,
            textEnd,
            textStart,
            visualEnd: visualIndex,
            visualStart: visualIndex - width,
            width,
        };

        if (isInHyperlink) {
            (segment as HyperlinkSegment).isHyperlink = true;
            (segment as HyperlinkSegment).hyperlinkUrl = currentUrl;
        }

        options.onSegment?.(segment, stateTracker);
    }

    // Handle any incomplete escape sequence
    if (escapeBuffer) {
        const segment: AnsiSegment = {
            isEscapeSequence: true,
            isGrapheme: false,
            text: escapeBuffer,
            textEnd: textIndex,
            textStart: textIndex - escapeBuffer.length,
            visualEnd: visualIndex,
            visualStart: visualIndex,
            width: 0,
        };

        options.onSegment?.(segment, stateTracker);
    }
};
