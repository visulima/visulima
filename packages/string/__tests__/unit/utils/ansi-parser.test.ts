import { describe, expect, it } from "vitest";

import { checkEscapeSequence, processAnsiString } from "../../../src/utils/ansi-parser";
import type { AnsiSegment, HyperlinkSegment } from "../../../src/utils/types";

describe("checkEscapeSequence", () => {
    it("should detect regular ANSI escape sequences", () => {
        expect.assertions(1);
        const text = "\u001B[31mHello";
        const chars = [...text];

        const result = checkEscapeSequence(chars, 0);

        expect(result).toStrictEqual({
            isInsideEscape: true,
            isInsideLinkEscape: false,
        });
    });

    it("should detect hyperlink start sequences", () => {
        expect.assertions(1);
        const text = "\u001B]8;;https://example.com\u0007Hello";
        const chars = [...text];

        const result = checkEscapeSequence(chars, 0);

        expect(result).toStrictEqual({
            isInsideEscape: true,
            isInsideLinkEscape: true,
        });
    });

    it("should return false for non-escape sequences", () => {
        expect.assertions(1);
        const text = "Hello World";
        const chars = [...text];

        const result = checkEscapeSequence(chars, 0);

        expect(result).toStrictEqual({
            isInsideEscape: false,
            isInsideLinkEscape: false,
        });
    });
});

describe("processAnsiString", () => {
    describe("regular ANSI sequences", () => {
        it("should process basic ANSI color codes", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B[31mHello\u001B[0m";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(7);
            expect(segments).toMatchSnapshot();
        });

        it("should handle nested ANSI sequences", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B[31m\u001B[1mBold Red\u001B[0m";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(11);
            expect(segments).toMatchSnapshot();
        });
    });

    describe("hyperlink handling", () => {
        it("should process hyperlink sequences", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B]8;;https://example.com\u0007Click here\u001B\\";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(12);
            expect(segments).toMatchSnapshot();
        });

        it("should handle nested ANSI sequences within hyperlinks", () => {
            expect.assertions(3);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B]8;;https://example.com\u0007\u001B[31mRed Link\u001B[0m\u001B\\";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(12);
            expect(segments.map((s) => s.text ?? "").join("")).toBe("\u001B[31mRed Link\u001B[0m")
            expect(segments).toMatchSnapshot();
        });

        it("should handle multiple hyperlinks", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B]8;;https://example1.com\u0007Link 1\u001B\\ and \u001B]8;;https://example2.com\u0007Link 2\u001B\\";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(21);
            expect(segments).toMatchSnapshot();
        });
    });

    describe("grapheme handling", () => {
        it("should handle combining characters", () => {
            expect.assertions(3);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "e\u0301"; // é using combining acute accent

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(2);
            expect(segments[0]).toMatchObject({
                isGrapheme: true,
                text: "e",
            });
            expect(segments[1]).toMatchObject({
                isGrapheme: true,
                text: "\u0301",
            });
        });

        it("should handle emoji sequences", () => {
            expect.assertions(8);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "👨‍👩‍👧‍👦"; // family emoji

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            // Each ZWJ-connected emoji component should be a separate segment
            expect(segments.length).toBeGreaterThan(1);
            segments.forEach((segment) => {
                expect(segment.isGrapheme).toBeTruthy();
            });
        });
    });

    describe("error handling", () => {
        it("should handle incomplete ANSI sequences", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B[31"; // Incomplete color code

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(1);
            expect(segments[0]).toMatchObject({
                isEscapeSequence: true,
                text: "\u001B[31",
            });
        });

        it("should handle incomplete hyperlink sequences", () => {
            expect.assertions(2);
            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B]8;;https://example.com"; // Missing BEL

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(1);
            expect(segments[0]).toMatchObject({
                hyperlinkUrl: "https://example.com",
                isEscapeSequence: true,
            });
        });
    });

    describe("callback control", () => {
        it("should stop processing when callback returns false", () => {
            expect.assertions(1);

            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "\u001B[31mHello\u001B[0m World";

            processAnsiString(text, {
                getWidth: (string_) => string_.length,
                onSegment: (segment) => {
                    segments.push(segment);
                    return segments.length < 2; // Stop after 2 segments
                },
            });

            expect(segments).toHaveLength(2);
        });

        it("should handle missing width function", () => {
            expect.assertions(2);

            const segments: (AnsiSegment | HyperlinkSegment)[] = [];
            const text = "Hello";

            processAnsiString(text, {
                onSegment: (segment) => {
                    segments.push(segment);
                    return true;
                },
            });

            expect(segments).toHaveLength(5);
            expect(segments[0].width).toBe(0);
        });
    });
});
