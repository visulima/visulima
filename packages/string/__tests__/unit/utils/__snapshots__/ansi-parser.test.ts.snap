// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processAnsiString > hyperlink handling > should handle multiple hyperlinks 1`] = `
[
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkStart": true,
    "width": 0,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "L",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "i",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "n",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "k",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": " ",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example1.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "1",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkEnd": true,
    "width": 0,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": " ",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "a",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "n",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "d",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": " ",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkStart": true,
    "width": 0,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "L",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "i",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "n",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "k",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": " ",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example2.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "2",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkEnd": true,
    "width": 0,
  },
]
`;

exports[`processAnsiString > hyperlink handling > should handle nested ANSI sequences within hyperlinks 1`] = `
[
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkStart": true,
    "width": 0,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[31m",
    "width": 0,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "R",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "e",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "d",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": " ",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "L",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "i",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "n",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "k",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[0m",
    "width": 0,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkEnd": true,
    "width": 0,
  },
]
`;

exports[`processAnsiString > hyperlink handling > should process hyperlink sequences 1`] = `
[
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkStart": true,
    "width": 0,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "C",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "l",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "i",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "c",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "k",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": " ",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "h",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "e",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "r",
    "width": 1,
  },
  {
    "hyperlinkUrl": "https://example.com",
    "isEscapeSequence": false,
    "isGrapheme": true,
    "isHyperlink": true,
    "text": "e",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "isHyperlink": true,
    "isHyperlinkEnd": true,
    "width": 0,
  },
]
`;

exports[`processAnsiString > regular ANSI sequences > should handle nested ANSI sequences 1`] = `
[
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[31m",
    "width": 0,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[1m",
    "width": 0,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "B",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "o",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "l",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "d",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": " ",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "R",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "e",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "d",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[0m",
    "width": 0,
  },
]
`;

exports[`processAnsiString > regular ANSI sequences > should process basic ANSI color codes 1`] = `
[
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[31m",
    "width": 0,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "H",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "e",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "l",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "l",
    "width": 1,
  },
  {
    "isEscapeSequence": false,
    "isGrapheme": true,
    "text": "o",
    "width": 1,
  },
  {
    "isEscapeSequence": true,
    "isGrapheme": false,
    "text": "[0m",
    "width": 0,
  },
]
`;
